os/binfmt/binfmt_execmodule.c

```diff
@@ -181,19 +181,19 @@ int exec_module(FAR struct binary_s *binp)
 	/* The first 4 bytes of the text section of the application must contain a
 	pointer to the application's mm_heap object. Here we will store the mm_heap
 	pointer to the start of the text section */
-	*(uint32_t *)(binp->alloc[ALLOC_DATA]) = (uint32_t)binp->uheap;
+	*(uint32_t *)(binp->datastart) = (uint32_t)binp->uheap;
 	rtcb = (struct tcb_s *)sched_self();
 	rtcb->uheap = (uint32_t)binp->uheap;
 
 	/* Initialize the MPU registers in tcb with suitable protection values */
 #ifdef CONFIG_ARM_MPU
 #ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
 	/* Configure text section as RO and executable region */
-	mpu_get_register_config_value(&rtcb->mpu_regs[3], MPU_REG_NUM_APP_TXT, (uintptr_t)binp->alloc[ALLOC_TEXT], binp->textsize, true, true);
+	mpu_get_register_config_value(&rtcb->mpu_regs[0], MPU_REG_NUM_APP_TXT, (uintptr_t)binp->alloc[ALLOC_TEXT], binp->textsize, true, true);
 	/* Configure ro section as RO and non-executable region */
-	mpu_get_register_config_value(&rtcb->mpu_regs[6], MPU_REG_NUM_APP_RO, (uintptr_t)binp->alloc[ALLOC_RO], binp->rosize, true, false);
+	mpu_get_register_config_value(&rtcb->mpu_regs[3], MPU_REG_NUM_APP_RO, (uintptr_t)binp->alloc[ALLOC_RO], binp->rosize, true, false);
 	/* Complete RAM partition will be configured as RW region */
-	mpu_get_register_config_value(&rtcb->mpu_regs[0], MPU_REG_NUM_APP_DATA, (uintptr_t)binp->alloc[ALLOC_DATA], binp->ramsize, false, false);
+	mpu_get_register_config_value(&rtcb->mpu_regs[6], MPU_REG_NUM_APP_DATA, (uintptr_t)binp->alloc[ALLOC_DATA], binp->ramsize, false, false);
```

Comment:
Could you let me know why this number change is necessary?

Reply:
This change is not related to the current issue. 
Actually, the ordering of the numbers does not matter. But the 0 -> 3 -> 6 order seems neat.

Reply:
We reorder the region by size for size optimization and usually data section is biggest because of heap. Didn't you change that to reduce the time of reordering as current??

Reply:
You are talking about the order in which the sections are allocated. That is different from the order in which the mpu regions are configured. The mpu regions can be configured in any order. It does not matter. It only matters if there are overlapping regions. Since, we dont have overlaps, its ok for us. In fact, the above change has nothing to do with any of the ordering. It just looks neat and makes more sense to set the values in numerical order, like 0 - 3- 6.. This is how I had implemented initially. But its got modified during some issue fix. So, I just corrected the order now

Reply:
ok